OneSignalSDK/onesignal/core/src/main/java/com/onesignal/internal/OneSignalImp.kt

```diff
@@ -179,169 +170,218 @@ internal class OneSignalImp : IOneSignal, IServiceProvider {
-                        pushSubscriptionModel.carrier = DeviceUtils.getCarrierName(
-                            services.getService<IApplicationService>().appContext,
-                        ) ?: ""
-                        pushSubscriptionModel.appVersion = AndroidUtils.getAppVersion(
-                            services.getService<IApplicationService>().appContext,
-                        ) ?: ""
-
-                        configModel!!.pushSubscriptionId = legacyPlayerId
-                        subscriptionModelStore!!.add(
-                            pushSubscriptionModel,
-                            ModelChangeTags.NO_PROPOGATE,
-                        )
-                        suppressBackendOperation = true
-                    }
+        // if an appId is supplied, we can move all loading in background and the countdown latch
+        // will free up accessors when initialization is fully completed
+        if (appId != null) {
+            OSPrimaryCoroutineScope.execute {
+                try {
+                    initEssentials(context)
 
-                    createAndSwitchToNewUser(suppressBackendOperation = suppressBackendOperation)
+                    if (!configModel!!.hasProperty(ConfigModel::appId.name) || configModel!!.appId != appId) {
+                        forceCreateUser = true
+                    }
 
-                    operationRepo!!.enqueue(
-                        LoginUserFromSubscriptionOperation(
-                            configModel!!.appId,
-                            identityModelStore!!.model.onesignalId,
-                            legacyPlayerId,
-                        ),
-                    )
-                    preferencesService!!.saveString(
-                        PreferenceStores.ONESIGNAL,
-                        PreferenceOneSignalKeys.PREFS_LEGACY_PLAYER_ID,
-                        null,
-                    )
+                    configModel!!.appId = appId
+                    updateConfig()
+                    val startupService = bootstrapServices()
+                    initUser(forceCreateUser)
+                    startupService.scheduleStart()
+                    isInitializedSuccess = true
+                } catch (e: Throwable) {
+                    Logging.error("initWithContext failed!", e)
+                } finally {
+                    initLatch.countDown()
                 }
-            } else {
-                Logging.debug("initWithContext: using cached user ${identityModelStore!!.model.onesignalId}")
             }
 
-            // schedule service starts out of main thread
-            startupService.scheduleStart()
-
             isInitialized = true
+
             return true
         }
+
+        // If the appId is not supplied, then it is called from internal classes. We will need to return
+        // a Boolean indicating whether it is safe to proceed.
+        initEssentials(context)
```

Comment:
This still does disk I/O on the main thread so it does look like this will solve ANRs that come from the FMC broadcast receiver. Can we also move this off the main thread?